### PR TITLE
Build riscv with c++ intrinsics

### PR DESCRIPTION
**Pull request by joy2myself**
This PR From GSoC Project "Optimize OpenCV for RISC-V"

This PR added cross compiling environment for RISC-V and use C++ version universal intrinsics for building.

<cut/>

PR contains:
Added toolchain file for target `riscv64-clang`.
Extended CMake for RISC-V and add instruction checks (cmake/checks/cpu_riscv.cpp).
Created intrin_riscv.hpp with C++ version universal intrinsics.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under OpenCV (BSD) License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
force_builders_only=docs,linux
```

### CODE REVIEW

File: cmake/OpenCVCompilerOptimizations.cmake [hunk 1]
```diff
@@ -102,6 +103,8 @@ ocv_optimization_process_obsolete_option(ENABLE_NEON NEON OFF)
 
 ocv_optimization_process_obsolete_option(ENABLE_VSX VSX ON)
 
+ocv_optimization_process_obsolete_option(ENABLE_RISCV RISCV OFF)
```
Comment (asmorkalov) [inline cmake/OpenCVCompilerOptimizations.cmake:106]: `ENABLE_RISCV` is bad choise as it's general architecture name. I propose to use something `ENABLE_RVV` or enable `ENABLE_RISCV_VEXT`.

File: cmake/checks/cpu_riscv.cpp [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+#if defined CV_RISCV
+int test()
+{
+    const float src[] = { 0.0f, 0.0f, 0.0f, 0.0f };
+    vfloat32m1_t val = vle32_v_f32m1((const float*)(src));
+    return (int)vfmv_f_s_f32m1_f32(val);
+}
+#else
+#error "RISCV is not supported"
```
Comment (asmorkalov) [inline cmake/checks/cpu_riscv.cpp:16]: "RISC-V vector extension"

**Comment by asmorkalov** [comment]
General recommendation: replace `RISCV` by `RISCV_RVV` or something associated with V-extension to not mix general features and vectorized code.

**Comment by asmorkalov** [comment]
There is build failure with clang:
```
/home/alexander/Projects/riscv/opencv/modules/core/src/system.cpp:607:37: error: use of undeclared identifier 'CV_CPU_RISCV'
        int baseline_features[] = { CV_CPU_BASELINE_FEATURES };
                                    ^
/home/alexander/Projects/riscv/opencv-build/cv_cpu_config.h:7:7: note: expanded from macro 'CV_CPU_BASELINE_FEATURES'
    , CV_CPU_RISCV \
      ^
/home/alexander/Projects/riscv/opencv/modules/core/src/system.cpp:608:53: error: invalid application of 'sizeof' to an incomplete type 'int []'
        if (!checkFeatures(baseline_features, sizeof(baseline_features) / sizeof(baseline_features[0]))
                                                    ^~~~~~~~~~~~~~~~~~~
/home/alexander/Projects/riscv/opencv/modules/core/src/system.cpp:619:52: error: invalid application of 'sizeof' to an incomplete type 'int []'
            checkFeatures(baseline_features, sizeof(baseline_features) / sizeof(baseline_features[0]), true);
                                                   ^~~~~~~~~~~~~~~~~~~
/home/alexander/Projects/riscv/opencv/modules/core/src/system.cpp:623:47: error: invalid application of 'sizeof' to an incomplete type 'int []'
        readSettings(baseline_features, sizeof(baseline_features) / sizeof(baseline_features[0]));
                                              ^~~~~~~~~~~~~~~~~~~
/home/alexander/Projects/riscv/opencv/modules/core/src/system.cpp:747:30: error: use of undeclared identifier 'CV_CPU_RISCV'
    const int features[] = { CV_CPU_BASELINE_FEATURES, CV_CPU_DISPATCH_FEATURES };
                             ^
/home/alexander/Projects/riscv/opencv-build/cv_cpu_config.h:7:7: note: expanded from macro 'CV_CPU_BASELINE_FEATURES'
    , CV_CPU_RISCV \
      ^
/home/alexander/Projects/riscv/opencv/modules/core/src/system.cpp:748:26: error: invalid application of 'sizeof' to an incomplete type 'const int []'
    const int sz = sizeof(features) / sizeof(features[0]);
```

**Comment by joy2myself** [comment]
I have updated the code for naming issue. Use `RVV` instead of `RISCV` now.

**Comment by asmorkalov** [comment]
There is build issue still:
```
[  6%] Building C object 3rdparty/libtiff/CMakeFiles/libtiff.dir/tif_aux.c.o
cd /home/alexander/Projects/riscv/opencv-build/3rdparty/libtiff && /home/alexander/Projects/riscv/rvv-llvm/build/bin/clang --target=riscv64-unknown-linux-gnu --sysroot=/home/alexander/Projects/riscv/riscv-gnu-toolchain-install/sysroot -DNEED_LIBPORT -D_FILE_OFFSET_BITS=64 -I/home/alexander/Projects/riscv/opencv-build/3rdparty/zlib -I/home/alexander/Projects/riscv/opencv/3rdparty/zlib -I/home/alexander/Projects/riscv/opencv-build/3rdparty/libtiff -I/home/alexander/Projects/riscv/opencv/3rdparty/libtiff -I/home/alexander/Projects/riscv/opencv-build/3rdparty/libjpeg-turbo -I/home/alexander/Projects/riscv/opencv/3rdparty/libjpeg-turbo/src -isystem /home/alexander/Projects/riscv/opencv-build  -march=rv64gcv --gcc-toolchain=/home/alexander/Projects/riscv/riscv-gnu-toolchain-install -w -march=rv64gcv --gcc-toolchain=/home/alexander/Projects/riscv/riscv-gnu-toolchain-install -w   -fsigned-char -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wstrict-prototypes -Winit-self -Wpointer-arith -Wsign-promo -Wuninitialized -Winit-self -Winconsistent-missing-override -Wno-delete-non-virtual-dtor -Wno-unnamed-type-template-args -Wno-comment -Wno-deprecated-enum-enum-conversion -Wno-deprecated-anon-enum-enum-conversion -fdiagnostics-show-option -pthread -Qunused-arguments -ffunction-sections -fdata-sections  -fvisibility=hidden -fvisibility-inlines-hidden -Wno-unused-but-set-variable -Wno-missing-prototypes -Wno-missing-declarations -Wno-undef -Wno-unused -Wno-sign-compare -Wno-cast-align -Wno-shadow -Wno-maybe-uninitialized -Wno-pointer-to-int-cast -Wno-int-to-pointer-cast -Wno-misleading-indentation -Wno-implicit-fallthrough -Wno-unused-parameter -O3 -DNDEBUG  -DNDEBUG -fPIC   -o CMakeFiles/libtiff.dir/tif_aux.c.o   -c /home/alexander/Projects/riscv/opencv/3rdparty/libtiff/tif_aux.c
In file included from /home/alexander/Projects/riscv/opencv/3rdparty/libtiff/tif_aux.c:30:
/home/alexander/Projects/riscv/opencv/3rdparty/libtiff/tiffiop.h:54:48: error: unknown type name 'size_t'
extern void *lfind(const void *, const void *, size_t *, size_t,
                                               ^
/home/alexander/Projects/riscv/opencv/3rdparty/libtiff/tiffiop.h:61:32: error: unknown type name 'size_t'
extern int snprintf(char* str, size_t size, const char* format, ...);
                               ^
2 errors generated.
```

**Comment by joy2myself** [comment]
I have updated the code. Fixed clang build issues.

**Comment by asmorkalov** [comment]
@joy2myself Please also take a look on CI status: https://pullrequest.opencv.org/buildbot/builders/precommit_docs/builds/25946

**Comment by joy2myself** [comment]
Hi Alex, I squashed commits to one and fixed the whitespace issue.

**Comment by asmorkalov** [comment]
Dockerfile with ready environment: https://github.com/asmorkalov/riscv-experiments/blob/master/Dockerfile
Docker image: https://hub.docker.com/repository/docker/asmorkalov/riscv_rvv_ci

File: modules/core/src/system.cpp [hunk 3]
```diff
@@ -597,6 +599,10 @@ struct HWFeatures
         have[CV_CPU_VSX3] = (CV_VSX3);
     #endif
 
+    #ifdef __riscv
+        have[CV_CPU_RVV] = true;
```
Comment (asmorkalov) [inline modules/core/src/system.cpp:603]: `__riscv` is very general and defined for general RISC-V architecture. There should be some specific check here.

File: modules/core/include/opencv2/core/hal/intrin.hpp [hunk 4]
```diff
@@ -199,9 +199,11 @@ using namespace CV_CPU_OPTIMIZATION_HAL_NAMESPACE;
 #   undef CV_VSX
 #   undef CV_FP16
 #   undef CV_MSA
+#   undef CV_RVV
 #endif
 
-#if (CV_SSE2 || CV_NEON || CV_VSX || CV_MSA || CV_WASM_SIMD) && !defined(CV_FORCE_SIMD128_CPP)
+
+#if (CV_SSE2 || CV_NEON || CV_VSX || CV_MSA || CV_WASM_SIMD/* || CV_RVV*/) && !defined(CV_FORCE_SIMD128_CPP)
```
Comment (asmorkalov) [inline modules/core/include/opencv2/core/hal/intrin.hpp:206]: Please revert the change here till we do not have specific optimizations.

File: platforms/linux/riscv64-clang.toolchain.cmake [hunk 5]
```diff
@@ -0,0 +1,27 @@
+set(CMAKE_SYSTEM_NAME Linux)
+set(CMAKE_SYSTEM_PROCESSOR riscv64)
+
+set(RISCV_CLANG_BUILD_ROOT /home/git/rvv-llvm/build CACHE PATH "Path to CLANG for RISC-V cross compiler build directory")
```
Comment (mshabunin) [inline platforms/linux/riscv64-clang.toolchain.cmake:4]: Maybe it would be better to use paths based in `/opt` as default locations?

File: modules/core/src/system.cpp [hunk 6]
```diff
@@ -597,6 +599,10 @@ struct HWFeatures
         have[CV_CPU_VSX3] = (CV_VSX3);
     #endif
 
+    #if defined __riscv && defined __riscv_vector
+        have[CV_CPU_RVV] = true;
```
Comment (mshabunin) [inline modules/core/src/system.cpp:603]: This check should be dynamic, otherwise dispatching does not make sense. Is it possible to detect vector extensions in runtime?

File: modules/core/include/opencv2/core/cv_cpu_dispatch.h [hunk 7]
```diff
@@ -168,6 +168,11 @@
 #  include <wasm_simd128.h>
 #endif
 
+#if defined(__riscv)
+#  define CV_RVV 1
+//#  include <riscv_vector.h>
```
[the remark this answers is not in the document]
Comment (joy2myself) [reply modules/core/include/opencv2/core/cv_cpu_dispatch.h:173]: This header is not needed now for C++ version universal intrinsics and will be needed later for rvv version universal intrinsics. Should I remove it now and add it later when it is needed?

File: platforms/linux/riscv64-clang.toolchain.cmake (hunk 5 above)
Comment (joy2myself) [reply platforms/linux/riscv64-clang.toolchain.cmake:4]: Now the release version of RISC-V toolchain does not support latest RISC-V vector extension. We use the latest version of toochain installed by code in GitHub. And Clang does not support also. We use a Github version that under development. Should I still edit the path here?

**Comment by joy2myself** [comment]
@mshabunin Hi, Maksim. Thanks a lot for your review. Firstly, I'd like to explain that this patch just added the toolchain file for cross compiling and extended some files for target RISC-V. The code in file `intrin_rvv.hpp` is all from the c++ version universal intrinsics. I'm not sure that part of the changes you requested about the file `intrin_rvv.hpp` were due to a misunderstanding of this or not.

File: modules/core/src/system.cpp (hunk 6 above)
Comment (joy2myself) [reply modules/core/src/system.cpp:603]: RISC-V vector extension spec have not mentioned about runtime check up to now. I will try to ask them about it.

File: platforms/linux/riscv64-clang.toolchain.cmake (hunk 5 above)
Comment (mshabunin) [reply platforms/linux/riscv64-clang.toolchain.cmake:4]: Are these paths hardcoded in the toolchain installation instruction?

Usually `/home` directory contains users' home directories, according to these lines your system have two users named `git` and `RISCV`. In some cases software can be installed in `/opt` directory, it is universal location and will not confuse readers.

File: modules/core/include/opencv2/core/cv_cpu_dispatch.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (mshabunin) [reply modules/core/include/opencv2/core/cv_cpu_dispatch.h:173]: Wouldn't it be included in `intrin_riscv.hpp`?

[the remark this answers is not in the document]
Comment (joy2myself) [reply modules/core/include/opencv2/core/cv_cpu_dispatch.h]: I think it might need to be included here. All the other implementation of universal intrinsics included their headers here but not in `intrin_XXX.hpp`. I haven't tested to included it in `intrin_rvv.hpp` but I think it should be consistent with other versions, should it? Then I just remove it at present.

File: platforms/linux/riscv64-clang.toolchain.cmake (hunk 5 above)
Comment (joy2myself) [reply platforms/linux/riscv64-clang.toolchain.cmake]: they are not hardcoded. It seems that my understanding of Linux path is not in place. Thanks for the guidance. I'll edit the default path here.

**Comment by joy2myself** [comment]
I have resolved changes except these in `intrin_rvv.cpp` and runtime check.

**Comment by joy2myself** [comment]
@mshabunin Hi, Maksim. I have finished all the changes you requested. Please review it again. Thank you!

File: modules/core/include/opencv2/core/cvdef.h [hunk 8]
```diff
@@ -269,6 +269,8 @@ namespace cv {
 #define CV_CPU_AVX512_CLX       261
 #define CV_CPU_AVX512_ICL       262
 
+#define CV_CPU_RVV              300
```
Comment (alalek) [inline modules/core/include/opencv2/core/cvdef.h:272]: Please use 210 here and below.

256+ is used for "groups".

Comment (joy2myself) [reply modules/core/include/opencv2/core/cvdef.h]: Fixed. Thanks a lot for your review!

**Comment by asmorkalov** [comment]
@alalek Can we merge the PR?
